### PR TITLE
Functions: Internals

### PR DESCRIPTION
**Pull request by octaharon**
Hereby implemented:
- framework internals functions
- flow control and conditional functions
- custom functons

Required implementation for JS/TS and Python

### CODE REVIEW

File: docs/syntax/140_functions.md [hunk 1]
```diff
@@ -72,10 +72,80 @@ contains(var, substring = 'searchString')
...
+yantrix codegen ./input.mermaid --functionFile functions.ts -o ./output.ts -c MyFSM -l typescript 
+```
+
+Obviously, injected functions must be implemented in target language. When trying to build from diagram that ncludes injected functions, which have not been provided, codegen will throw a build-time error. 
+
+Typically a function file should contain a dictionary with named functions, stored as first-class citizens in a given language. When building for language, that does not support storing functions in object keys, modularization tecnhiques should be used. JS/TS users can benefit from both worlds and, even more so, they can import built-in functions directly into their custom implementations:
+
+```typescript
+// functions.ts
+import {coalesce} from '@yantrix/functions`
```
Comment (octaharon) [inline docs/syntax/140_functions.md]: typo

File: docs/syntax/140_functions.md [hunk 2]
```diff
@@ -118,10 +118,80 @@ contains(var, substring = 'searchString')
...
+define/<FUNCTION_NAME> (<ARGUMENTS_LIST>) => <RETURN_VALUE>
+```
+
+Regardless of whether it's a `Transformer` or a `Predicate`, this function can be used anywhere in the same diagram or a `Slice`, including "before" the definition in the source code 
+
+Custom function can reference any type of functions, including itself, other inline functions, built-ins and injected functions. They are calculated just-in-time and exhibit little difference to built-ins
+
+### Injecting functions
+
+Sometimes native syntax is just not enough. In this case you can implement certain functions in a target programming language (or few) or even use existing APIs in your system to plug the business logic into diagrams, allowing for fast and easy [integrations](../integrations/).
```
Comment (obl1v1onZzz) [inline docs/syntax/140_functions.md]: Duplicate description

**Review by obl1v1onZzz — CHANGES_REQUESTED** [review]

File: packages/functions/src/internals.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+ *
+ * @template T - The type of the automata.
+ *
+ * @param _ - The constructor of the automata.
+ * @returns A function that takes an automata instance and returns its current state ID or null.
+ */
+function currentStateId<T extends GenericAutomata>(_: new (...args: any[]) => T): (automata: T) => number | null {
+	return (automata: T) => automata.state;
+}
+/**
```
Comment (obl1v1onZzz) [inline packages/functions/src/internals.ts]: I think all functions that accept the automaton should do a check of instanceOf

**Review by obl1v1onZzz — APPROVED** [review]

File: packages/codegen/src/core/modules/JavaScript/JavaScriptCompiler/dictionaries/serializer.ts [hunk 4]
```diff
@@ -122,6 +122,35 @@ export function getActionToStateFromState(props: {
...
+	lines.push(`\t"currentActionId": internalFunctions.currentActionId(${className}),`);
+	lines.push(`\t"currentActionName": internalFunctions.currentActionName(${className}, actionsDictionary),`);
+	lines.push(`\t"currentCycle": internalFunctions.currentCycle(${className}),`);
+	lines.push(`\t"currentEpoch": internalFunctions.currentEpoch(epoch),`);
+	lines.push(`}`);
+	return lines.join('\n');
+}
+
+export function getAutomataEpochCounterCode() {
+	return `const epoch = { val: 1 };`;
```
Comment (octaharon) [inline packages/codegen/src/core/modules/JavaScript/JavaScriptCompiler/dictionaries/serializer.ts]: turn into wrapper function `incrementEpoch()`
add `getEpoch()` 
move to global namespace or shared module, maybe along with EventBus
call `incrementEpoch` wherever, do not access the object directly

File: packages/functions/src/internals.ts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+ * @param object - An object where keys are items and values are their weights.
+ * @returns A randomly selected value based on the weights.
+ * @throws An error if the object contains NaN values or if no value can be selected.
+ */
+function weightedRandom(inputObject: { [key: string]: number }): string {
+	// https://trekhleb.medium.com/weighted-random-in-javascript-4748ab3a1500
+
+	// Check if any of the object's values is NaN or negative
+	for (const value of Object.values(inputObject)) {
+		if (Number.isNaN(value)) throw new Error('Weighted random object contains NaN values');
```
Comment (octaharon) [inline packages/functions/src/internals.ts]: the value should pass if `Number.isInteger` and `>0`

File: packages/functions/src/internals.ts [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+		else if (value < 0) throw new Error('Weighted random object contains negative values');
+	}
+
+	// remove items with weights of 0 from object
+	const object = Object.fromEntries(Object.entries(inputObject).filter(([_, value]) => value > 0));
```
Comment (octaharon) [inline packages/functions/src/internals.ts]: the filter is redundant if weight=0 throws an error

File: packages/functions/src/internals.ts [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+	const objectKeys: string[] = Object.keys(object);
+	const objectValues: number[] = Object.values(object);
```
Comment (octaharon) [inline packages/functions/src/internals.ts]: change to type inference

File: packages/functions/src/internals.ts [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+	// prepare the list of cumulative weights
+	const cumulativeWeights: number[] = [];
+	for (let i = 0; i < objectValues.length; i++) {
+		cumulativeWeights.push(objectValues[i]! + (cumulativeWeights[i - 1] || 0));
```
Comment (octaharon) [inline packages/functions/src/internals.ts]: 1) calculate sum of all weights
2) make a local accumulator variable, divide it by sum each time  and add directly to `cumulativeWeights`

File: packages/functions/src/internals.ts [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	// random number from 0 to the highest cumulative weight value
+	const randomNumber = Math.floor(Math.random() * cumulativeWeights[cumulativeWeights.length - 1]!);
```
Comment (octaharon) [inline packages/functions/src/internals.ts]: roll basic random()

File: packages/functions/__tests__/internals.test.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+	it('should handle equal weights', () => {
+		for (let i = 0; i < iterations; i++) {
+			const weight = randomInteger(1, 100);
+			const weights = { a: weight, b: weight, c: weight };
+			const result = weightedRandom(weights);
+			expect(['a', 'b', 'c']).toContain(result);
+		}
+	});
+
+	it('should handle zero weights correctly', () => {
```
Comment (octaharon) [inline packages/functions/__tests__/internals.test.ts]: should throw with zero weights

File: packages/functions/__tests__/internals.test.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+			const weights = { a: randomInteger(1, 100) };
+			const result = weightedRandom(weights);
+			expect(result).toBe('a');
+		}
+	});
+
+	it('should handle equal weights', () => {
+		for (let i = 0; i < iterations; i++) {
+			const weight = randomInteger(1, 100);
+			const weights = { a: weight, b: weight, c: weight };
```
Comment (octaharon) [inline packages/functions/__tests__/internals.test.ts]: use the same key array to construct object  `weights`, that is later used to check for inclusion of keys

File: packages/functions/__tests__/internals.test.ts [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+	});
+
+	it('should handle zero weights correctly', () => {
+		for (let i = 0; i < iterations; i++) {
+			const weights = { a: 0, b: 0, c: randomInteger(1, 100) };
+			const result = weightedRandom(weights);
+			expect(result).toBe('c');
+		}
+	});
+
```
Comment (octaharon) [inline packages/functions/__tests__/internals.test.ts:95]: add stochastic tests, i.e.
100x `weightedRandom({a:3,b:1})` should contain 75+-1 `a` and 25+-` b`

File: packages/functions/src/conditionals.ts [hunk 13]
```diff
@@ -169,21 +169,13 @@ const coalesce = variadic(<T>(values: (T | null | undefined)[]): T | null => {
...
- *
- * @param min - The minimum value (inclusive) for the random number. If not provided, defaults to 0 or 1.
- * @param max - The maximum value (exclusive) for the random number. If not provided, defaults to 0 or 1.
- * @returns A random number between min and max, or 0 or 1 if min and max are not provided.
- */
-function random(min?: number, max?: number): number {
-	if (isNumber(min) && isNumber(max)) return Math.floor(Math.random() * (max - min) + min);
-	else return Math.round(Math.random());
+function choose(index: number, ...options: any[]): any {
+	return options[index];
```
Comment (octaharon) [inline packages/functions/src/conditionals.ts]: must throw if index is invalid, <0 or is greater than amount of options

File: packages/functions/__tests__/conditionals.test.ts [hunk 14]
```diff
@@ -69,34 +70,29 @@ describe('function Tests', () => {
 		});
 	});
 
-	describe('random Function', () => {
-		const iterations = 1000;
+	describe('choose Function', () => {
```
Comment (octaharon) [inline packages/functions/__tests__/conditionals.test.ts:73]: add a couple of well-defined tests, i.e.
`choose([100,200,300],1)`===200

add tests that check for throwing when index is out of range or invalid

File: packages/functions/__tests__/conditionals.test.ts [hunk 15]
```diff
@@ -69,34 +70,29 @@ describe('function Tests', () => {
...
-		it(`should handle edge cases with negative ranges (${iterations} iterations)`, () => {
-			const min = -10;
-			const max = -5;
-			for (let i = 0; i < iterations; i++) {
-				const result = builtInFunctions.random(min, max);
-				expect(result).toBeGreaterThanOrEqual(min);
-				expect(result).toBeLessThan(max);
-			}
+		it('should return undefined when no options are provided', () => {
+			expect(choose(0)).toBeUndefined();
```
Comment (octaharon) [inline packages/functions/__tests__/conditionals.test.ts]: must throw an error

File: docs/syntax/140_functions.md [hunk 16]
```diff
@@ -118,10 +121,23 @@ contains(var, substring = 'searchString')
...
-| `random`    |        (**Number**, **Number**) => **Number**         |                                                                                     | a uniform random number between the first and the second arguments                                       |
+| `if`        |             (**Binary**, any, any) => any             | - Condition<br/>- Yes option<br>- No option| `Yes option`, if `Condition` is truthy, or `No option` otherwise |
+| `case`      | (**Binary**, any, [**Binary**, any], ..., any) => any | - Condition 1<br/>- Option 1<br/>- Condition 2<br/>- Option 2</br>- ...<br/>- Default  | `Option N`, if `Condition N` is truthy, or `Default` if none are |
+| `coalesce`  |                   (any, ..) => any                    | any collection of `Expressions` | first non-Null value in the list of arguments|
+| `choose`    |                   (**Number**, [any,...] => any         | - Index<br/>- Option 1<br/>- Option 2<br/>- ...   | picks `Index+1`th option from arguments, i.e. `choose(0,"a","b")` returns `"a"` |
```
Comment (octaharon) [inline docs/syntax/140_functions.md]: update to reflect changes:
- can't be called with only 1 argument
- Index must be within bounds

File: packages/functions/src/internals.ts [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+function currentEpoch(epochRef: { val: number }): () => { val: number } {
+	return () => epochRef;
+}
+/**
+ * Retrieves the current timestamp in milliseconds.
+ *
+ * @returns The current timestamp in milliseconds since the Unix epoch.
+ */
+function currentTimestamp(): number {
+	return Date.now(); // in milliseconds, for microseconds need process.hrtime() from node.js
```
Comment (octaharon) [inline packages/functions/src/internals.ts]: - should be isomorphic
- use `process.hrtime` is available, `Math.floor(performance.now()*1000)` else, or `Date.now()*1000` if none of the above are applicable
